docs/u2f_migration.md

```diff
@@ -0,0 +1,157 @@
+      COSE::Key::EC2.new(
+        alg: COSE::Algorithm.by_name("ES256").id,
+        crv: 1,
+        x: decoded_public_key[1..32],
+        y: decoded_public_key[33..-1]
+      )
+    else
+      raise "expected U2F public key to be in uncompressed point format"
+    end
+  end
+end
+```
+
+Of course a `verify` instance method is not implemented, as there is no real interaction with an interaction going on.
+Like the readme prescribes, you need to store at least the object's `credential.id` and `credential.public_key` in your
+database. 
+
+For reasons explained further below, it is also recommended to store a boolean value whether or not a stored credential
+if migrated from U2F or not. For this documented, we'll assume this database column is called `u2f_migrated`.
+
+```ruby
+migrated_credential_response = U2fCredentialMigrator.new(
+  domain: domain,
+  u2f_certificate: u2f_registration.certificate,
+  u2f_key_handle: u2f_registration.key_handle,
+  u2f_public_key: u2f_registration.public_key,
+  u2f_counter: u2f_registration.counter
+)
+migrated_credential_response.credential.id
+# => "\x99\xB5LE83I>q.\xE9\x9C\x90l\xED'\xD5E[\xAB\xDE9\xB7\xCD!\x85\x92\x9F{\x13\xA8\x86"
+migrated_credential_response.credential.public_key
+# => "\xA5\x03& \x01!X \xE2P^Q`\xF9\x97\xD9*n<\x14\xDA\xB6a\xEEoK\x03\xACpMb\xED\x8B\x06E\"#!\xED\xC6\x01\x02\"X #C\x97\xAD C\x000\xE7\xD1\xD4%\xCFh\x83\xCD\x9E\xCB\xBC,\"\x1F>\xF6SZ\xA1U\xAB7\xBE\xEB"
+``` 
+
+## Authenticate migrated U2F credentials
+
+Following the gem's readme on the [authentication initiation](https://github.com/cedarcode/webauthn-ruby/blob/master/README.md#authentication), 
+we need to add the [FIDO AppID extension](https://www.w3.org/TR/webauthn/#sctn-appid-extension):
+
+```ruby
+credential_request_options = WebAuthn.credential_request_options
+credential_request_options[:allowCredentials] = current_user.webauthn_credentials.map do |credential|
+  { id: credential.id, type: "public-key" }
+end
+if current_user.webauthn_credentials.any?(&:u2f_migrated)
+  credential_request_options[:extensions] = { appId: domain.to_s }
+end
+```
+
+During authentication verification phase, we must make a change for U2F migrated credentials that use the AppID extension.
+The WebAuthn specification explains:
+
+> If true, the AppID was used and thus, when verifying an assertion, the Relying Party MUST expect the `rpIdHash` to be the hash of the _AppID_, not the [RP ID](https://www.w3.org/TR/webauthn/#rp-id). 
+
+Concretely:
+- AppID: https://login.example.com
+- Valid RP IDs: login.example.com (default) and example.com
+
+```ruby
+assertion_response = WebAuthn::AuthenticatorAssertionResponse.new(
+  authenticator_data: authenticator_data,
+  client_data_json: client_data_json,
+  signature: signature
+)
+
+assertion_response.verify(expected_challenge, allowed_credentials: [allowed_credential], rp_id: domain.to_s)
+```
+
+Note that this will break if WebAuthn registered credentials are used.
```

Comment:
If the goal is for a site to move entirely to move to webauthn, doesn't that make this bad advice?

Reply:
Yes, been playing with https://github.com/cedarcode/webauthn-ruby/pull/211 to solve this but I felt it was important to point out it's not completely solved yet.